File: app/src/main/java/com/example/swob_deku/Models/RMQ/RMQConnectionService.java
package com.example.swob_deku.Models.RMQ;

import static com.example.swob_deku.GatewayClientListingActivity.GATEWAY_CLIENT_LISTENERS;

import android.app.Notification;
import android.app.PendingIntent;
import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.os.IBinder;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;

import com.example.swob_deku.BroadcastReceivers.IncomingTextSMSBroadcastReceiver;
import com.example.swob_deku.BroadcastReceivers.IncomingTextSMSReplyActionBroadcastReceiver;
import com.example.swob_deku.Commons.Helpers;
import com.example.swob_deku.GatewayClientListingActivity;
import com.example.swob_deku.Models.GatewayClients.GatewayClient;
import com.example.swob_deku.Models.GatewayClients.GatewayClientHandler;
import com.example.swob_deku.Models.SMS.SMSHandler;
import com.example.swob_deku.R;
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import com.rabbitmq.client.DeliverCallback;
import com.rabbitmq.client.Delivery;
import com.rabbitmq.client.RecoveryDelayHandler;
import com.rabbitmq.client.ShutdownListener;
import com.rabbitmq.client.ShutdownSignalException;
import com.rabbitmq.client.impl.DefaultExceptionHandler;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeoutException;

public class RMQConnectionService extends Service {
    final int NOTIFICATION_ID = 1234;
    final long DELAY_TIMEOUT = 10000;

    public final static String RMQ_SUCCESS_BROADCAST_INTENT = "RMQ_SUCCESS_BROADCAST_INTENT";
    public final static String RMQ_STOP_BROADCAST_INTENT = "RMQ_STOP_BROADCAST_INTENT";

    private HashMap<Integer, RMQMonitor> connectionList = new HashMap<>();

    private ExecutorService consumerExecutorService;

    private BroadcastReceiver messageStateChangedBroadcast;

    private HashMap<Long, Map<Long, Channel>> channelList = new HashMap<>();

    private SharedPreferences sharedPreferences;

    private SharedPreferences.OnSharedPreferenceChangeListener sharedPreferenceChangeListener;

    @Override
    public void onCreate() {
        super.onCreate();

        consumerExecutorService = Executors.newFixedThreadPool(5); // Create a pool of 5 worker threads
        handleBroadcast();

        sharedPreferences = getSharedPreferences(GATEWAY_CLIENT_LISTENERS, Context.MODE_PRIVATE);

        registerListeners();

    }

    public int[] getGatewayClientNumbers() {
        Map<String, ?> keys = sharedPreferences.getAll();
        int running = 0;
        int reconnecting = 0;

        for(String _key : keys.keySet()) {
            if (sharedPreferences.getBoolean(_key, false))
                ++running;
            else
                ++reconnecting;
        }

        return new int[]{running, reconnecting};
    }

   private void registerListeners() {
       sharedPreferenceChangeListener = new SharedPreferences.OnSharedPreferenceChangeListener() {
           @Override
           public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
               Log.d(getClass().getName(), "Shared preferences changed: " + key);
               if(connectionList.containsKey(Integer.parseInt(key))) {
                   if(connectionList.get(Integer.parseInt(key)) != null &&
                           !sharedPreferences.contains(key) ) {
                       new Thread(new Runnable() {
                           @Override
                           public void run() {
                               try {
                                   stop(Integer.parseInt(key));
                               } catch (Exception e) {
                                   e.printStackTrace();
                               }
                           }
                       }).start();
                   } else if(connectionList.get(Integer.parseInt(key)) != null &&
                           sharedPreferences.contains(key) ){
                       // This means something connected
                       int[] states = getGatewayClientNumbers();
                       createForegroundNotification(states[0], states[1]);
                   }
//                   else if(connectionList.get(Integer.parseInt(key)) != null &&
//                           sharedPreferences.contains(key) && !sharedPreferences.getBoolean(key, false)) {
//                       // This means something connected
//                       createForegroundNotification(--runningGatewayClientCount,
//                               ++reconnectingGatewayClientCount);
//                   }
               }
               else {
                   new Thread(new Runnable() {
                       @Override
                       public void run() {
                           GatewayClientHandler gatewayClientHandler = new GatewayClientHandler(getApplicationContext());
                           try {
                               GatewayClient gatewayClient = gatewayClientHandler.fetch(Integer.parseInt(key));
                               connectGatewayClient(gatewayClient);
                           } catch (InterruptedException e) {
                               e.printStackTrace();
                           }
                       }
                   }).start();
               }
           }
       };
       sharedPreferences.registerOnSharedPreferenceChangeListener(sharedPreferenceChangeListener);
   }

    private void handleBroadcast() {
        messageStateChangedBroadcast = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, @NonNull Intent intent) {
                // TODO: in case this intent comes back but the internet connection broke to send back acknowledgement
                // TODO: should store pending confirmations in a place
                if (intent.hasExtra(IncomingTextSMSReplyActionBroadcastReceiver.BROADCAST_STATE)) {
                    if (intent.getStringExtra(IncomingTextSMSReplyActionBroadcastReceiver.BROADCAST_STATE)
                            .equals(IncomingTextSMSReplyActionBroadcastReceiver.SENT_BROADCAST_INTENT) &&
                            intent.hasExtra(RMQConnection.MESSAGE_GLOBAL_MESSAGE_ID_KEY)) {
                        Log.d(getClass().getName(), "Service received a broadcast and should acknowledge to rmq from here");

                        long globalMessageId = intent.getLongExtra(RMQConnection.MESSAGE_GLOBAL_MESSAGE_ID_KEY, -1);
                        if (globalMessageId != -1) {
                            Map<Long, Channel> deliveryChannel = channelList.get(globalMessageId);
                            Long deliveryTag = deliveryChannel.keySet().iterator().next();
                            Channel channel = deliveryChannel.get(deliveryTag);
                            if (channel != null && channel.isOpen()) {
                                new Thread(new Runnable() {
                                    @Override
                                    public void run() {
                                        try {
                                            channel.basicAck(deliveryTag, false);
                                        } catch (IOException e) {
                                            e.printStackTrace();
                                        }
                                    }
                                }).start();
                            }
                        }
                    } else if (intent.getStringExtra(IncomingTextSMSReplyActionBroadcastReceiver.BROADCAST_STATE)
                            .equals(IncomingTextSMSReplyActionBroadcastReceiver.FAILED_BROADCAST_INTENT) &&
                            intent.hasExtra(RMQConnection.MESSAGE_GLOBAL_MESSAGE_ID_KEY)) {

                        long globalMessageId = intent.getLongExtra(RMQConnection.MESSAGE_GLOBAL_MESSAGE_ID_KEY, -1);
                        if (globalMessageId != -1) {
                            Map<Long, Channel> deliveryChannel = channelList.get(globalMessageId);
                            Long deliveryTag = deliveryChannel.keySet().iterator().next();
                            Channel channel = deliveryChannel.get(deliveryTag);
                            if (channel != null && channel.isOpen()) {
                                new Thread(new Runnable() {
                                    @Override
                                    public void run() {
                                        try {
                                            channel.basicReject(deliveryTag, true);
                                        } catch (IOException e) {
                                            e.printStackTrace();
                                        }
                                    }
                                }).start();
                            }
                        }
                    }
                }
            }
        };

        registerReceiver(messageStateChangedBroadcast,
                new IntentFilter(SMSHandler.MESSAGE_STATE_CHANGED_BROADCAST_INTENT));
    }

    private DeliverCallback getDeliverCallback(RMQConnection rmqConnection) {
        return new DeliverCallback() {
            @Override
            public void handle(String consumerTag, Delivery delivery) throws IOException {
                String message = new String(delivery.getBody(), StandardCharsets.UTF_8);
                try {
                    JSONObject jsonObject = new JSONObject(message);

                    String body = jsonObject.getString(RMQConnection.MESSAGE_BODY_KEY);
                    Log.d(getClass().getName(), "New request body: " + body);

                    String msisdn = jsonObject.getString(RMQConnection.MESSAGE_MSISDN_KEY);
                    String globalMessageKey = jsonObject.getString(RMQConnection.MESSAGE_GLOBAL_MESSAGE_ID_KEY);

                    long messageId = Helpers.generateRandomNumber();

                    PendingIntent[] pendingIntents = IncomingTextSMSBroadcastReceiver
                            .getPendingIntentsForServerRequest(getApplicationContext(), messageId,
                                    Long.parseLong(globalMessageKey));

                    // TODO: fix subscriptionId to actually be the value
                    SMSHandler.sendTextSMS(getApplicationContext(), msisdn, body,
                            pendingIntents[0], pendingIntents[1], messageId, null);

                    Map<Long, Channel> deliveryChannelMap = new HashMap<>();
                    deliveryChannelMap.put(delivery.getEnvelope().getDeliveryTag(), rmqConnection.getChannel());

                    channelList.put(Long.valueOf(globalMessageKey), deliveryChannelMap);
                } catch (JSONException e) {
                    e.printStackTrace();
                    rmqConnection.getChannel().basicReject(delivery.getEnvelope().getDeliveryTag(), false);
                } catch(Exception e) {
                    e.printStackTrace();
                }
            }
        };
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.d(getClass().getName(), "Request to start service received...");
        Map<String, ?> storedGatewayClients = sharedPreferences.getAll();
        GatewayClientHandler gatewayClientHandler = new GatewayClientHandler(getApplicationContext());
        for (String gatewayClientIds : storedGatewayClients.keySet()) {
            try {
                GatewayClient gatewayClient = gatewayClientHandler.fetch(Integer.parseInt(gatewayClientIds));
                Log.d(getClass().getName(), "* Starting new RMQ connection: " + gatewayClient.getFriendlyConnectionName());
                connectGatewayClient(gatewayClient);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        return START_STICKY;
    }

    public void connectGatewayClient(GatewayClient gatewayClient) throws InterruptedException {
        if(!connectionList.containsKey(gatewayClient.getId())) {
            Log.d(getClass().getName(), "Starting new service connection...");
            ConnectionFactory factory = new ConnectionFactory();

            factory.setRecoveryDelayHandler(new RecoveryDelayHandler() {
                @Override
                public long getDelay(int recoveryAttempts) {
                    // TODO: send notification informing reconnecting is being attempted
                    Log.d(getClass().getName(), "Attempting auto recovery...");

                    connectionList.get(gatewayClient.getId()).setConnected(DELAY_TIMEOUT);
                    Log.d(getClass().getName(), "Done setting reconnect");
                    return DELAY_TIMEOUT;
                }
            });

            factory.setUsername(gatewayClient.getUsername());
            factory.setPassword(gatewayClient.getPassword());
            factory.setVirtualHost(gatewayClient.getVirtualHost());
            factory.setHost(gatewayClient.getHostUrl());
            factory.setPort(gatewayClient.getPort());
            factory.setConnectionTimeout(15000);
            factory.setExceptionHandler(new DefaultExceptionHandler());

            // TODO: create a full handler to manage the retry to connection
            // TODO: which matches the Android WorkManager methods
            Thread thread = new Thread(new Runnable() {
                @Override
                public void run() {
                    try {
                        /**
                         * Avoid risk of :ForegroundServiceDidNotStartInTimeException
                         * - Put RMQ connection in list before connecting which could take a while
                         */

                        RMQConnection rmqConnection = new RMQConnection();

                        RMQMonitor rmqMonitor = new RMQMonitor(getApplicationContext(),
                                gatewayClient.getId(),
                                rmqConnection);
                        connectionList.put(gatewayClient.getId(), rmqMonitor);

                        rmqMonitor.setConnected(DELAY_TIMEOUT);

                        Connection connection = factory.newConnection(consumerExecutorService,
                                gatewayClient.getFriendlyConnectionName());

                        rmqMonitor.setConnected(0L);
                        connection.addShutdownListener(new ShutdownListener() {
                            @Override
                            public void shutdownCompleted(ShutdownSignalException cause) {
                                Log.d(getClass().getName(), "Connection shutdown cause: " + cause.toString());
                            }
                        });
                        rmqMonitor.getRmqConnection().setConnection(connection);

                        if(gatewayClient.getProjectName() != null && !gatewayClient.getProjectName().isEmpty()) {
                            rmqConnection.createQueue(gatewayClient.getProjectName(),
                                    gatewayClient.getProjectBinding(), getDeliverCallback(rmqConnection));
                            rmqConnection.consume();
                        }

//                        sharedPreferences.edit().putBoolean(String.valueOf(gatewayClient.getId()), true)
//                                        .apply();

//                        createForegroundNotification(++runningGatewayClientCount,
//                                reconnectingGatewayClientCount);

                    } catch (IOException | TimeoutException e) {
                        e.printStackTrace();
                        // TODO: send a notification indicating this, with options to retry the connection
                        int[] states = getGatewayClientNumbers();
                        createForegroundNotification(states[0], states[1]);
                    }
                }
            });
            thread.start();
        }
    }

    private void stop(int gatewayClientId) {
        try {
            if(connectionList.containsKey(gatewayClientId)) {
                connectionList.remove(gatewayClientId)
                        .getRmqConnection().close();
                if(connectionList.isEmpty()) {
                    stopForeground(true);
                    stopSelf();
                }
                else {
                    int[] states = getGatewayClientNumbers();
                    createForegroundNotification(states[0], states[1]);
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        Log.d(getClass().getName(), "Ending connection...");
        if(messageStateChangedBroadcast != null)
            unregisterReceiver(messageStateChangedBroadcast);
        sharedPreferences.unregisterOnSharedPreferenceChangeListener(sharedPreferenceChangeListener);
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    private void createForegroundNotification(int runningGatewayClientCount, int reconnecting) {
        Intent notificationIntent = new Intent(getApplicationContext(), GatewayClientListingActivity.class);
        PendingIntent pendingIntent =
                PendingIntent.getActivity(getApplicationContext(), 0, notificationIntent,
                        PendingIntent.FLAG_IMMUTABLE);

        String description = runningGatewayClientCount + " " + getString(R.string.gateway_client_running_description);

        if(reconnecting > 0)
            description += "\n" + reconnecting + " " + getString(R.string.gateway_client_reconnecting_description);

        Notification notification =
                new NotificationCompat.Builder(getApplicationContext(), getString(R.string.running_gateway_clients_channel_id))
                        .setContentTitle(getString(R.string.gateway_client_running_title))
                        .setSmallIcon(R.drawable.ic_stat_name)
                        .setPriority(NotificationCompat.DEFAULT_ALL)
                        .setSilent(true)
                        .setOngoing(true)
                        .setContentText(description)
                        .setContentIntent(pendingIntent)
                        .build();

        startForeground(NOTIFICATION_ID, notification);
    }
}

File: app/src/main/java/com/afkanerd/deku/QueueListener/RMQ/RMQConnectionService.kt
package com.afkanerd.deku.QueueListener.RMQ

import android.app.Activity
import android.app.PendingIntent
import android.app.Service
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.content.pm.ServiceInfo
import android.os.Build
import android.os.Bundle
import android.os.IBinder
import android.provider.Telephony
import android.telephony.SubscriptionInfo
import android.util.Log
import android.view.inputmethod.CorrectionInfo
import androidx.core.app.NotificationCompat
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import com.afkanerd.deku.DefaultSMS.BroadcastReceivers.IncomingTextSMSBroadcastReceiver
import com.afkanerd.deku.DefaultSMS.Models.Conversations.Conversation
import com.afkanerd.deku.Datastore
import com.afkanerd.deku.DefaultSMS.Models.NativeSMSDB
import com.afkanerd.deku.DefaultSMS.Models.SIMHandler
import com.afkanerd.deku.DefaultSMS.Models.SMSDatabaseWrapper
import com.afkanerd.deku.DefaultSMS.R
import com.afkanerd.deku.Modules.SemaphoreManager
import com.afkanerd.deku.Modules.ThreadingPoolExecutor
import com.afkanerd.deku.QueueListener.GatewayClients.GatewayClient
import com.afkanerd.deku.QueueListener.GatewayClients.GatewayClientHandler
import com.afkanerd.deku.QueueListener.GatewayClients.GatewayClientListingActivity
import com.afkanerd.deku.QueueListener.GatewayClients.GatewayClientProjects
import com.rabbitmq.client.Channel
import com.rabbitmq.client.ConnectionFactory
import com.rabbitmq.client.ConsumerShutdownSignalCallback
import com.rabbitmq.client.DeliverCallback
import com.rabbitmq.client.Delivery
import com.rabbitmq.client.ShutdownSignalException
import com.rabbitmq.client.impl.DefaultExceptionHandler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json
import org.json.JSONException
import org.junit.Assert
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.util.concurrent.TimeoutException

class RMQConnectionService : Service() {
    private lateinit var databaseConnector: Datastore
    private lateinit var subscriptionInfoList: List<SubscriptionInfo>
    private lateinit var messageStateChangedBroadcast: BroadcastReceiver
    private var gatewayClientLiveData: LiveData<GatewayClient> = MutableLiveData()

    val observer = Observer<GatewayClient> {
        Log.d(javaClass.name, "Gateway client changed")
        if(!it.activated) {
            stopSelf()
        }
        else if(it.state == GatewayClient.STATE_DISCONNECTED)
            ThreadingPoolExecutor.executorService.execute {
                connectGatewayClient(it)
            }
    }
    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        createForegroundNotification()
        val gatewayClientId = intent.getLongExtra(GatewayClient.GATEWAY_CLIENT_ID, -1)
        Assert.assertTrue(gatewayClientId.toInt() != -1)
        gatewayClientLiveData = databaseConnector.gatewayClientDAO().fetchLiveData(gatewayClientId)
        gatewayClientLiveData.observeForever(observer)
//        ThreadingPoolExecutor.executorService.execute {
//            val gatewayClient = databaseConnector.gatewayClientDAO().fetch(gatewayClientId)
//            connectGatewayClient(gatewayClient)
//        }
//        gatewayClientLiveData.value!!.state = GatewayClient.STATE_RECONNECTING
//        databaseConnector.gatewayClientDAO().update(gatewayClientLiveData.value)
        return START_STICKY
    }

    override fun onDestroy() {
        unregisterReceiver(messageStateChangedBroadcast)
        gatewayClientLiveData.removeObserver(observer)
        ThreadingPoolExecutor.executorService.execute {
            val gatewayClient = gatewayClientLiveData.value
            gatewayClient?.state = GatewayClient.STATE_DISCONNECTED
            Datastore.getDatastore(applicationContext).gatewayClientDAO().update(gatewayClient)
        }
        super.onDestroy()
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    override fun onCreate() {
        super.onCreate()
        databaseConnector = Datastore.getDatastore(applicationContext)
        subscriptionInfoList = SIMHandler.getSimCardInformation(applicationContext)
        handleBroadcast()
    }

    private fun handleBroadcast() {
        val intentFilter = IntentFilter()
        intentFilter.addAction(IncomingTextSMSBroadcastReceiver.SMS_SENT_BROADCAST_INTENT)
        messageStateChangedBroadcast = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                if (intent.action != null && intentFilter.hasAction(intent.action)) {
                    Log.d(javaClass.name, "Received incoming broadcast")
                    if (intent.hasExtra(RMQConnection.MESSAGE_SID) &&
                            intent.hasExtra(RMQConnection.RMQ_DELIVERY_TAG)) {

                        val sid = intent.getStringExtra(RMQConnection.MESSAGE_SID)
                        val messageId = intent.getStringExtra(NativeSMSDB.ID)

                        val consumerTag = intent.getStringExtra(RMQConnection.RMQ_CONSUMER_TAG)
                        val deliveryTag =
                                intent.getLongExtra(RMQConnection.RMQ_DELIVERY_TAG, -1)
                        val rmqID = intent.getLongExtra(RMQConnection.RMQ_ID, -1)

                        Assert.assertTrue(!consumerTag.isNullOrEmpty())
                        Assert.assertTrue(deliveryTag != -1L)
                        Assert.assertTrue(rmqID != -1L)

                        rmqConnectionList.forEach {rmq ->
                            if(rmq.id == rmqID) {
                                rmq.findChannelByTag(consumerTag!!)?.let {
                                    Log.d(javaClass.name, "Received an ACK of the message...")
                                    if (resultCode == Activity.RESULT_OK) {
                                        ThreadingPoolExecutor.executorService.execute {
                                            if (it.isOpen) it.basicAck(deliveryTag, false)
                                        }
                                    } else {
                                        Log.w(javaClass.name, "Rejecting message sent")
                                        ThreadingPoolExecutor.executorService.execute {
                                            if (it.isOpen) it.basicReject(deliveryTag, true)
                                        }
                                    }
                                }
                            }
                        }

                    }
                }
            }
        }

        registerReceiver(messageStateChangedBroadcast, intentFilter, RECEIVER_EXPORTED)
    }

    @Serializable
    private data class SMSRequest(val text: String, val to: String, val sid: String, val id: String)
    private suspend fun sendSMS(smsRequest: SMSRequest,
                                subscriptionId: Int,
                                consumerTag: String,
                                deliveryTag: Long,
                                rmqConnectionId: Long) {
        SemaphoreManager.resourceSemaphore.acquire()
        val messageId = System.currentTimeMillis()
        SemaphoreManager.resourceSemaphore.release()

        val threadId = Telephony.Threads.getOrCreateThreadId(applicationContext, smsRequest.to)

        val bundle = Bundle()
        bundle.putString(RMQConnection.MESSAGE_SID, smsRequest.sid)
        bundle.putString(RMQConnection.RMQ_CONSUMER_TAG, consumerTag)
        bundle.putLong(RMQConnection.RMQ_DELIVERY_TAG, deliveryTag)
        bundle.putLong(RMQConnection.RMQ_ID, rmqConnectionId)

        val conversation = Conversation()
        conversation.message_id = messageId.toString()
        conversation.text = smsRequest.text
        conversation.address = smsRequest.to
        conversation.subscription_id = subscriptionId
        conversation.type = Telephony.Sms.MESSAGE_TYPE_OUTBOX
        conversation.date = System.currentTimeMillis().toString()
        conversation.thread_id = threadId.toString()
        conversation.status = Telephony.Sms.STATUS_PENDING

        databaseConnector.threadedConversationsDao()
                .insertThreadAndConversation(applicationContext, conversation)
        SMSDatabaseWrapper.send_text(applicationContext, conversation, bundle)
        Log.d(javaClass.name, "SMS sent...")
    }
    private fun getDeliverCallback(channel: Channel, subscriptionId: Int,
                                   rmqConnectionId: Long): DeliverCallback {
        return DeliverCallback { consumerTag: String, delivery: Delivery ->
            val message = String(delivery.body, StandardCharsets.UTF_8)

            CoroutineScope(Dispatchers.IO).launch {
                try {
                    val smsRequest = Json.decodeFromString<SMSRequest>(message)
                    sendSMS(smsRequest,
                            subscriptionId,
                            consumerTag,
                            delivery.envelope.deliveryTag,
                            rmqConnectionId)
                } catch (e: Exception) {
                    Log.e(javaClass.name, "", e)
                    when(e) {
                        is SerializationException -> {
                            channel.let {
                                if (it.isOpen)
                                    it.basicReject(delivery.envelope.deliveryTag, false)
                            }
                        }
                        is IllegalArgumentException -> {
                            channel.let {
                                if (it.isOpen)
                                    it.basicReject(delivery.envelope.deliveryTag, true)
                            }
                        }
                        else -> {
                            e.printStackTrace()
                        }
                    }
                }
            }
        }
    }


    private val rmqConnectionList = ArrayList<RMQConnection>()
    private fun startConnection(factory: ConnectionFactory, gatewayClient: GatewayClient) {
        Log.d(javaClass.name, "Starting new connection...")

        // TODO: use this without internet connection and catch that error and best handle
        gatewayClient.state = GatewayClient.STATE_INITIALIZING
        try {
            val connection = factory.newConnection(ThreadingPoolExecutor.executorService,
                    gatewayClient.friendlyConnectionName)
            gatewayClient.state = GatewayClient.STATE_CONNECTED
            databaseConnector.gatewayClientDAO().update(gatewayClient)

            val rmqConnection = RMQConnection(gatewayClient.id, connection)
            rmqConnectionList.add(rmqConnection)

            connection.addShutdownListener {
                /**
                 * The logic here, if the user has not deactivated this - which can be known
                 * from the database connection state then reconnect this client
                 */
                Log.e(javaClass.name, "Connection shutdown cause: $it")
                val gatewayClient1 = databaseConnector.gatewayClientDAO().fetch(gatewayClient.id)
                if(gatewayClient1.activated) {
                    rmqConnectionList.remove(rmqConnection)
                    gatewayClient1.state = GatewayClient.STATE_RECONNECTING
                    databaseConnector.gatewayClientDAO().update(gatewayClient1)
                    GatewayClientHandler.startListening(applicationContext, gatewayClient)
                }
            }

            val gatewayClientProjectsList = databaseConnector.gatewayClientProjectDao()
                    .fetchGatewayClientIdList(gatewayClient.id)

            for (i in gatewayClientProjectsList.indices) {
                for (j in subscriptionInfoList.indices) {
                    val channel = rmqConnection.createChannel()
                    val gatewayClientProjects = gatewayClientProjectsList[i]
                    val subscriptionId = subscriptionInfoList[j].subscriptionId
                    val bindingName = if (j > 0)
                        gatewayClientProjects.binding2Name else gatewayClientProjects.binding1Name

                    Log.d(javaClass.name, "Starting channel for sim slot $j in project #$i")
                    startChannelConsumption(rmqConnection, channel, subscriptionId,
                            gatewayClientProjects, bindingName)
                }
            }

        } catch (e: Exception) {
            when(e) {
                is TimeoutException -> {
                    e.printStackTrace()
                    Thread.sleep(3000)
                    Log.d(javaClass.name, "Attempting a reconnect to the server...")
                    startConnection(factory, gatewayClient)
                }
                is IOException -> {
                    Log.e(javaClass.name, "IO Exception connecting rmq", e)
                }
                else -> {
                    Log.e(javaClass.name, "Exception connecting rmq", e)
                }
            }
        }
    }

    private fun startChannelConsumption(rmqConnection: RMQConnection,
                                channel: Channel,
                                subscriptionId: Int,
                                gatewayClientProjects: GatewayClientProjects,
                                bindingName: String) {
        Log.d(javaClass.name, "Starting channel connection")
        channel.basicRecover(true)
        val deliverCallback = getDeliverCallback(channel, subscriptionId, rmqConnection.id)
        val queueName = rmqConnection.createQueue(gatewayClientProjects.name, bindingName, channel)
        val messagesCount = channel.messageCount(queueName)

        val consumerTag = channel.basicConsume(queueName, false, deliverCallback,
                object : ConsumerShutdownSignalCallback {
                    override fun handleShutdownSignal(consumerTag: String, sig: ShutdownSignalException) {
                        if (rmqConnection.connection.isOpen) {
                            Log.e(javaClass.name, "Consumer error", sig)
                            startChannelConsumption(rmqConnection,
                                    rmqConnection.createChannel(),
                                    subscriptionId,
                                    gatewayClientProjects,
                                    bindingName)
                        }
                    }
                })
        Log.d(javaClass.name, "Created Queue: $queueName ($messagesCount) - tag: $consumerTag")
        rmqConnection.bindChannelToTag(channel, consumerTag)
    }

    private fun connectGatewayClient(gatewayClient: GatewayClient) {
        Log.d(javaClass.name, "Starting new service connection...")

        val factory = ConnectionFactory()
        factory.username = gatewayClient.username
        factory.password = gatewayClient.password
        factory.virtualHost = gatewayClient.virtualHost
        factory.host = gatewayClient.hostUrl
        factory.port = gatewayClient.port
        factory.isAutomaticRecoveryEnabled = true
        factory.exceptionHandler = DefaultExceptionHandler()

        factory.setRecoveryDelayHandler {
            Log.w(javaClass.name, "Factory recovering...: $it")
            10000
        }

        startConnection(factory, gatewayClient)
    }


    private fun stop(gatewayClientId: Long) {
        val iterator = rmqConnectionList.iterator()
        while(iterator.hasNext()) {
            val gatewayClient = iterator.next()
            if(gatewayClient.id == gatewayClientId) {
                rmqConnectionList.remove(gatewayClient)
                break
            }
        }

    }

    private fun createForegroundNotification() {
        val notificationIntent = Intent(applicationContext, GatewayClientListingActivity::class.java)
        val pendingIntent = PendingIntent
                .getActivity(applicationContext,
                        0,
                        notificationIntent,
                        PendingIntent.FLAG_IMMUTABLE)

//        databaseConnector.gatewayClientDAO().fetch()
//
//        var nConnected = 0
//        var nReconnecting = 0
//        gatewayClients.forEach {
//            if(it.activated) {
//                when(it.state) {
//                    GatewayClient.STATE_CONNECTED -> nConnected++
//                    GatewayClient.STATE_RECONNECTING -> nReconnecting++
//                }
//                if (nConnected < 1 && nReconnecting < 1) {
//                    stopForeground(STOP_FOREGROUND_REMOVE)
//                    stopSelf()
//                }
//            }
//        }

//        val description = "$nConnected ${getString(R.string.gateway_client_running_description)}\n" +
//                "$nReconnecting ${getString(R.string.gateway_client_reconnecting_description)}"
        val description = applicationContext.getString(R.string.gateway_client_running_description)

        val notification =
                NotificationCompat.Builder(applicationContext,
                        applicationContext.getString(R.string.running_gateway_clients_channel_id))
                        .setContentTitle(applicationContext
                                .getString(R.string.gateway_client_running_title))
                        .setSmallIcon(R.drawable.ic_stat_name)
                        .setPriority(NotificationCompat.DEFAULT_ALL)
                        .setSilent(true)
                        .setOngoing(true)
                        .setContentText(description)
                        .setContentIntent(pendingIntent)
                        .build()

        val NOTIFICATION_ID = 1234
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
            startForeground(NOTIFICATION_ID, notification,
                    ServiceInfo.FOREGROUND_SERVICE_TYPE_DATA_SYNC)
        } else startForeground(NOTIFICATION_ID, notification)
    }
}
